core/src/layers/madsim.rs

```diff
@@ -0,0 +1,316 @@
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn blocking_write(
+        &self,
+        path: &str,
+        args: OpWrite,
+    ) -> crate::Result<(RpWrite, Self::BlockingWriter)> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn blocking_list(
+        &self,
+        path: &str,
+        args: OpList,
+    ) -> crate::Result<(RpList, Self::BlockingPager)> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn blocking_scan(
+        &self,
+        path: &str,
+        args: OpScan,
+    ) -> crate::Result<(RpScan, Self::BlockingPager)> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+}
+
+pub struct MadsimReader {
+    path: String,
+    args: OpRead,
+    handle: ServiceNodeHandle,
+}
+
+impl oio::Read for MadsimReader {
+    fn poll_read(&mut self, _cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<crate::Result<usize>> {
+        let mut inner = self.handle.inner.lock().unwrap();
+        if let Some((_meta, data)) = inner.get(&self.path) {
+            buf.copy_from_slice(data);
+            Poll::Ready(Ok(data.len()))
+        } else {
+            Poll::Ready(Err(crate::Error::new(
+                crate::ErrorKind::NotFound,
+                "not found",
+            )))
+        }
+    }
+
+    fn poll_seek(&mut self, cx: &mut Context<'_>, pos: SeekFrom) -> Poll<crate::Result<u64>> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Option<crate::Result<Bytes>>> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+}
+
+impl oio::BlockingRead for MadsimReader {
+    fn read(&mut self, buf: &mut [u8]) -> crate::Result<usize> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn seek(&mut self, pos: SeekFrom) -> crate::Result<u64> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn next(&mut self) -> Option<crate::Result<Bytes>> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+}
+
+pub struct MadsimWriter {
+    path: String,
+    args: OpWrite,
+    handle: ServiceNodeHandle,
+}
+
+impl oio::BlockingWrite for MadsimWriter {
+    fn write(&mut self, bs: Bytes) -> crate::Result<()> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn append(&mut self, bs: Bytes) -> crate::Result<()> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+
+    fn close(&mut self) -> crate::Result<()> {
+        unimplemented!("currently not supported, will be implemented in the future")
+    }
+}
+
+#[async_trait]
+impl oio::Write for MadsimWriter {
```

Comment:
Just curious why the `Write` trait is async-trait style while the `Read` trait is poll style. 😄 @Xuanwo

Reply:
`reqwest` takes our input as a whole and there is no place for us to implement `poll_write` :rofl: